### PR TITLE
Issue 4977: Add Actions CI for Gradle, Codecov, Java 11

### LINKED ISSUE #4977: Codecov bash uploader frequently fails in Travis CI
**Problem description**
Codecov bash uploader is not fault tolerant, and therefore Codecov status checks are sometimes unreliable. For example, here is a connection time-out when uploading to codecov.io: https://travis-ci.org/github/pravega/pravega/jobs/711340911#L7103 (click on line 7103 to expose line 7104 for the error message)

The Codecov Action appears to be more reliable than the bash uploader that lacks any kind of resilience.

**Problem location**
PRs occasionally fail to validate themselves due to issues downloading the Codecov bash script.

**Suggestions for an improvement**
Migrate to GitHub Actions. GitHub Actions has also been experiencing intermittent issues, but this was recently diagnosed by them and there are known workarounds (that is, use checkout@v1 action instead of checkout@v2 action). See this comment and the one before it: https://github.com/actions/checkout/issues/237#issuecomment-662662682

I have set up `pravega-cdi` with GitHub Actions as an example: https://github.com/derekm/pravega-cdi/pull/1

### PR DESCRIPTION
**Pull request by derekm**
**Change log description**  
Adds build and test/coverage for Java 11. Runs Rat, Checkstyle and SpotBugs before tests.

**Purpose of the change**  
To test the Gradle build against Java 11. To more reliably upload coverage reports to Codecov. To transition off of Travis CI. To run source checks prior to tests.
Fixes #4977

**What the code does**  
Adds a workflow to run compilation against a matrix of Java versions (currently only 11) and tests against a single Java version (11) on a simulated merge.

**How to verify it**  
Actions CI is running in this PR. It also runs in my fork here: https://github.com/derekm/pravega/pull/4

**Open questions**
1. Should this incorporate release or system test stages from the Travis build?

### CODE REVIEW

**Comment by eolivelli** [comment]
@derekm in my opinion you have to enable "GitHub Actions" for this repository.
In my experience once you create those files the PR is automatically evaluated by GitHub.

Regarding CodeCov I think that we should use only one JDK as reference, otherwise we will have too much spam

**Comment by derekm** [comment]
@eolivelli -- It looks like @fpj enabled Actions for us!

> Regarding CodeCov I think that we should use only one JDK as reference, otherwise we will have too much spam

It is probably best to restrict the codecov upload to whatever version of Java we build releases with. The existing Travis CI runs two builds, one for the branch and one for the merge with master. Both builds upload their coverage results, so whichever build finishes last affects the PR most.

Should we upload codecov for the branch builds or the merge builds?

**Comment by eolivelli** [comment]
I would say the merge build.

**Comment by eolivelli** [comment]
I remember another awkward fact with github  actions.
Only owners of the repo are allowed to restart failed jobs.
With Travis the owner of the branch can do it.
In BK repository we added a custom github action that listens for comments and in case of a magic word it triggers the build

**Comment by derekm** [comment]
> I remember another awkward fact with github actions.
> Only owners of the repo are allowed to restart failed jobs.
> With Travis the owner of the branch can do it.
> In BK repository we added a custom github action that listens for comments and in case of a magic word it triggers the build

@eolivelli, I have been able to restart all jobs in this PR. Notice it is building now because I restarted it. See the Checks tab for the previous success and the present re-run.

That's a cool idea for a github bot!

File: .github/workflows/gradle.yml [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        key: gradle-m2-java-8
+        restore-keys: |
+          gradle-m2-java-
+          gradle-m2
+    - name: Set up JDK 8
+      uses: actions/setup-java@v1
+      with:
+        java-version: 8
+    - name: Check with Gradle
+      run: ./gradlew check
```
Comment (tkaitchuck) [inline .github/workflows/gradle.yml:67]: If we want this to be fast and not run integration tests, just add the argument:
`--exclude-task test:integration:test`

Comment (andreipaduroiu) [reply .github/workflows/gradle.yml:67]: I think we do want to run integration tests. Maybe not for all JDKs, but at least for one of them.

**Comment by andreipaduroiu** [comment]
@derekm Are we still pursuing this?

**Comment by derekm** [comment]
> @derekm Are we still pursuing this?

@andreipaduroiu -- Not actively. I think as a Proof-of-Concept, this branch/PR is sufficient.

We need to make a decision about whether this exploration provides any value and whether we want to utilize it.

Now that we're up to Java 11 for most of the code, there may be further changes needed here (besides turning on integration tests for the preferred JDK version) if we decide to adopt this approach.

I can get this branch up-to-date with master so we can make final assessments.

**Comment by fpj** [comment]
@derekm the last comment does not make clear how we are supposed to come to a closure on this. what is preventing us from coming to a closure on this PR?

**Comment by derekm** [comment]
@fpj -- I will carve out some time to get this PR updated with the latest master branch & mixed Java 8 + Java 11 build procedure. I'm not sure who makes the decision that GitHub Actions is preferred to Travis CI, but getting back to being equivalent to our Travis build is key to moving forward with this. If we merge this, we can run for a time with Actions and Travis in parallel.

**Comment by derekm** [comment]
This PR depends on PR #5402. When PR #5402 is merged, the merge of `andreipaduroiu/issue-5401-leaks` into this branch will need to be undone, and this branch will need to be rebased onto `master`.

File: .github/workflows/gradle.yml (hunk 1 above)
Comment (derekm) [reply .github/workflows/gradle.yml:67]: `./gradlew assemble` is run against Java 11 and Java 14 in parallel, and `./gradlew check` (including integration tests) is run against only Java 11 after the assemble workflow is successful.

**Comment by derekm** [comment]
This branch is ready for merge consideration. Codecov says the `codecov/patch` and `codecov/project` results are awaiting completion of all build steps.

**Review by andreipaduroiu — APPROVED** [review]

**Review by tkaitchuck — APPROVED** [review]

**Comment by RaulGracia** [comment]
@derekm @medvedevigorek @fpj the Git build is failing consistently due to tests failures. I think that we need to look into that before merging this PR, otherwise we will switch to a broken build.

**Comment by andreipaduroiu** [comment]
@derekm Let's try the steps suggested by @shrids . They should improve the build and check style times. They do not affect unit tests and since we get 2 CPUs on these build boxes, we should theoretically see some improvement there.

**Comment by RaulGracia** [comment]
@derekm @medvedevigorek @andreipaduroiu @tkaitchuck the failing CLI tests are fixed in this PR, but the build is failing very often (more than on Travis) on various tests. If we merge this one without having consistent green builds with Git Actions, we would be at risk of blocking merges to master.

File: .github/workflows/gradle.yml [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+          gradle-m2-java-
+          gradle-m2
+    - name: Set up JDK ${{ matrix.java }}
+      uses: actions/setup-java@v1
+      with:
+        java-version: ${{ matrix.java }}
+    - name: Assemble with Gradle
+      run: ./gradlew assemble
+    - name: Javadoc with Gradle
+      run: ./gradlew javadocs
```
[the remark this answers is not in the document]
Comment (derekm) [reply .github/workflows/gradle.yml]: thanks

File: .github/workflows/gradle.yml [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        key: gradle-m2-java-${{ matrix.java }}
+        restore-keys: |
+          gradle-m2-java-
+          gradle-m2
+    - name: Set up JDK ${{ matrix.java }}
+      uses: actions/setup-java@v1
+      with:
+        java-version: ${{ matrix.java }}
+    - name: Assemble with Gradle
+      run: ./gradlew assemble
```
[the remark this answers is not in the document]
Comment (derekm) [reply .github/workflows/gradle.yml]: with only 2 CPUs, the reduction in runtime isn't as dramatic; but if we ever get to targeting our own hosted runners (perhaps via CNCF infra), maybe we can allocate more CPUs later on

**Comment by derekm** [comment]
Moving from Java 8 to Java 11 greatly increases the rates of test failure in the GitHub Actions build environment.

Something about Java 11 required upping Gradle's 512 MB default heap size for Test tasks in order to fix a class of random OOM failures. Some examples of these failures include:

```
io.pravega.cli.user.stream.StreamCommandTest$SecureStreamCommandsTest > testAppendAndReadStream FAILED
    java.lang.OutOfMemoryError: Java heap space
        at io.pravega.common.util.CircularBuffer.<init>(CircularBuffer.java:33)
        at io.pravega.client.segment.impl.SegmentInputStreamImpl.<init>(SegmentInputStreamImpl.java:71)
        at io.pravega.client.segment.impl.SegmentInputStreamFactoryImpl.getEventSegmentReader(SegmentInputStreamFactoryImpl.java:64)
        at io.pravega.client.segment.impl.SegmentInputStreamFactoryImpl.getEventSegmentReader(SegmentInputStreamFactoryImpl.java:58)
        at io.pravega.client.segment.impl.SegmentInputStreamFactoryImpl.createEventReaderForSegment(SegmentInputStreamFactoryImpl.java:48)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.acquireSegmentsIfNeeded(EventStreamReaderImpl.java:348)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.updateGroupStateIfNeeded(EventStreamReaderImpl.java:292)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEventInternal(EventStreamReaderImpl.java:140)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEvent(EventStreamReaderImpl.java:125)
        at io.pravega.cli.user.stream.StreamCommand$Read.execute(StreamCommand.java:260)
        at io.pravega.cli.user.TestUtils.executeCommand(TestUtils.java:42)
        at io.pravega.cli.user.stream.StreamCommandTest.testAppendAndReadStream(StreamCommandTest.java:122)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
        at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.lang.Thread.run(Thread.java:834)


io.pravega.cli.user.stream.StreamCommandTest$SecureStreamCommandsTest > testAppendAndReadStream FAILED
    java.util.concurrent.CompletionException: java.lang.OutOfMemoryError: Java heap space
        at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
        at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
        at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1702)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)

        Caused by:
        java.lang.OutOfMemoryError: Java heap space
            at io.netty.util.internal.PlatformDependent.allocateUninitializedArray(PlatformDependent.java:281)
            at io.netty.buffer.UnpooledUnsafeHeapByteBuf.allocateArray(UnpooledUnsafeHeapByteBuf.java:39)
            at io.netty.buffer.UnpooledByteBufAllocator$InstrumentedUnpooledUnsafeHeapByteBuf.allocateArray(UnpooledByteBufAllocator.java:144)
            at io.netty.buffer.UnpooledHeapByteBuf.<init>(UnpooledHeapByteBuf.java:59)
            at io.netty.buffer.UnpooledUnsafeHeapByteBuf.<init>(UnpooledUnsafeHeapByteBuf.java:34)
            at io.netty.buffer.UnpooledByteBufAllocator$InstrumentedUnpooledUnsafeHeapByteBuf.<init>(UnpooledByteBufAllocator.java:139)
            at io.netty.buffer.UnpooledByteBufAllocator.newHeapBuffer(UnpooledByteBufAllocator.java:82)
            at io.netty.buffer.AbstractByteBufAllocator.heapBuffer(AbstractByteBufAllocator.java:168)
            at io.netty.buffer.AbstractByteBufAllocator.heapBuffer(AbstractByteBufAllocator.java:159)
            at io.netty.buffer.Unpooled.buffer(Unpooled.java:118)
            at io.pravega.client.connection.impl.CommandEncoder.<init>(CommandEncoder.java:78)
            at io.pravega.client.connection.impl.TcpClientConnection.lambda$connect$1(TcpClientConnection.java:200)
            at io.pravega.client.connection.impl.TcpClientConnection$$Lambda$570/0x00000001005ce840.get(Unknown Source)
            at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
            ... 6 more


io.pravega.cli.user.stream.StreamCommandTest > testAppendAndReadStream FAILED
    java.lang.OutOfMemoryError: Java heap space
        at io.pravega.common.util.CircularBuffer.<init>(CircularBuffer.java:33)
        at io.pravega.client.segment.impl.SegmentInputStreamImpl.<init>(SegmentInputStreamImpl.java:71)
        at io.pravega.client.segment.impl.SegmentInputStreamFactoryImpl.getEventSegmentReader(SegmentInputStreamFactoryImpl.java:64)
        at io.pravega.client.segment.impl.SegmentInputStreamFactoryImpl.getEventSegmentReader(SegmentInputStreamFactoryImpl.java:58)
        at io.pravega.client.segment.impl.SegmentInputStreamFactoryImpl.createEventReaderForSegment(SegmentInputStreamFactoryImpl.java:48)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.acquireSegmentsIfNeeded(EventStreamReaderImpl.java:348)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.updateGroupStateIfNeeded(EventStreamReaderImpl.java:292)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEventInternal(EventStreamReaderImpl.java:140)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEvent(EventStreamReaderImpl.java:125)
        at io.pravega.cli.user.stream.StreamCommand$Read.execute(StreamCommand.java:260)
        at io.pravega.cli.user.TestUtils.executeCommand(TestUtils.java:42)
        at io.pravega.cli.user.stream.StreamCommandTest.testAppendAndReadStream(StreamCommandTest.java:117)
```

Alleviating the tight heap space problem (which is [known to be an issue in the new G1 GC](https://docs.oracle.com/en/java/javase/11/gctuning/garbage-first-garbage-collector.html#GUID-BE157AF6-29E7-461A-82CF-50C1978785DA)) by giving Gradle Test tasks 2GB of heap exposes other classes of random test failures, which have been known, but that are exacerbated by the combination of constrained Actions build environment and Java 11.

**Comment by derekm** [comment]
As per issue #4998, intermittent test failures has been an ongoing issue, and success rates have worsened. Building the 0.8 release branch with both Java 8 and Java 11 suggests that worsening success rates could be related to using Java 11 (as I observe more difficulty validating the 0.8.1 release on Java 11 with a successful run of the unit and integration tests).

Fixing flakey/flaky tests that are prone to intermittent failures will be needed for reliable Actions and Travis builds in any case, but with default settings in Java 11, Travis is presently more reliable than Actions. We need to end-of-life our Travis builds to avoid a paid plan.

It has been decided to temporarily pursue a paid Travis plan to get reliable test runs while we complete a program of fixing flaky tests. We can run Actions & Travis builds side-by-side while keeping the Actions checks optional and the Travis checks required, and as we complete fixing tests, the Actions build will begin to pass, and we can end-of-life the Travis build. Therefore, it has been decided to merge this PR and make the Actions build a required check at a later time while we continue relying on Travis for passing builds (it is expected Travis will produce more successful builds if a paid plan offers build VMs with more resources).

I will use a separate branch & PR to continue investigating Java 11 environmental fixes that might restore the former reliability of the Actions build (which was previously at least on par with the free Travis builds).

Any final comments about the merging of this PR?

**Comment by andreipaduroiu** [comment]
@derekm I am planning to merge this to master today (when it passes the Travis build). We will then work out the GH Actions failures later, but as long as GH Actions is not mandatory, we should be OK. 

We need this in so we can validate our flaky test fixes in other branches.

**Comment by derekm** [comment]
@andreipaduroiu -- It looks like we will have a passing build here soon. It is running the Javadocs job here: https://travis-ci.org/github/pravega/pravega/builds/753577142

Following @shrids advice, which I confirmed, we are able to remove the `javadocs` run if there is a successful `assemble` run. The Javadocs run in Travis isn't a no-op, since Travis runs the command on a new machine. Is the goal in Travis to run the javadocs step in all cases in case the assemble step fails? Or can the javadocs run be removed from Travis also?

**Comment by andreipaduroiu** [comment]
I wouldn't touch Travis at this point.

**Pull request merged**